data_steward/cdr_cleaner/cleaning_rules/base_cleaning_rule.py

```diff
@@ -0,0 +1,248 @@
+"""
+A base class for cleaning rules to extend and implement.
+
+DC-718
+
+This is a base class that all cleaning rules should extend.  This base class
+should contain data relevant to each cleaning rule.  It's attributes and
+functions should be used to ensure that new cleaning rules are added with
+an adequate amount of information.  This is useful for things like, reporting
+features.
+"""
+# Python imports
+import logging
+from abc import ABC, abstractmethod
+from typing import List
+
+# Third party imports
+import googleapiclient
+import oauth2client
+
+# Project imports
+import constants.cdr_cleaner.clean_cdr as cdr_consts
+
+LOGGER = logging.getLogger(__name__)
+
+
+class AbstractBaseCleaningRule(ABC):
+    """
+    Contains attributes and functions relevant to all cleaning rules.
+    """
+    string_list = List[str]
+
+    def __init__(self):
+        """
+        Instantiate a cleaning rule with basic attributes.
+        """
+        super().__init__()
+
+    @abstractmethod
+    def get_query_dictionary_list(self, *args, **keyword_args):
+        """
+        Interface to return a list of query dictionaries.
+
+        :returns:  a list of query dictionaries
+        """
+        pass
+
+    @abstractmethod
+    def print_queries(self, *args, **keyword_args):
+        """
+        Helper function to print the SQL a class geenerates.
+        """
+        pass
+
+
+class BaseCleaningRule(AbstractBaseCleaningRule):
+    """
+    Contains attributes and functions relevant to all cleaning rules.
+    """
+    string_list = List[str]
+    cleaning_class_list = List[AbstractBaseCleaningRule]
+
+    def __init__(self,
+                 jira_issue_numbers: string_list = None,
+                 description: str = None,
+                 affected_datasets: string_list = None,
+                 project_id: str = None,
+                 dataset_id: str = None,
+                 sandbox_dataset_id: str = None,
+                 depends_on: cleaning_class_list = None):
+        """
+        Instantiate a cleaning rule with basic attributes.
+
+        Inheriting classes must set issue numbers, description and affected
+        datasets.  As other tickets may affect the SQL of a cleaning rule,
+        append them to the list of Jira Issues.
+        DO NOT REMOVE ORIGINAL JIRA ISSUE NUMBERS!
+
+        :param jira_issue_numbers:  a list of strings inidcating jira issues
```

Comment:
Interesting idea.  Since this is something that we will need to be able to build reports off of, would it make sense to add `issue_url` as an additional parameter rather than remove `jira_issue_numbers`?  Also, possibly rename `jira_issue_numbers` to just `issue_numbers`.  Thoughts?

Reply:
@mark-velez Do we want our Jira urls stored in the repository?

Reply:
Sorry, I missed your first response. Good point, there may be reasons not to have the urls at this time. I think having just `issue_numbers` is fine.

Reply:
@mark-velez So, `issue_urls` is in or out?  The current patch has it in.  Someone would still have to have an approved jira account to access the issues.  When we decide this, I can push this in.

Reply:
@mark-velez I've set issue_urls to default to an empty list.  I doubt anyone will ever set it if it isn't required.  We can revisit the default later if needed.